controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -135,72 +129,30 @@ public static ObjectTypeAttributeDefinition getAttributeDefinition() {
     public static ObjectTypeAttributeDefinition.Builder getAttributeBuilder(String name, String xmlName, boolean allowNull) {
         return new ObjectTypeAttributeDefinition.Builder(name, credentialStoreAttribute, credentialAliasAttribute, credentialTypeAttribute, clearTextAttribute)
                 .setXmlName(xmlName)
-                .setAttributeMarshaller(credentialReferenceAttributeMarshaller())
-                .setAttributeParser(credentialReferenceAttributeParser())
+                .setAttributeMarshaller(AttributeMarshaller.ATTRIBUTE_OBJECT)
+                .setAttributeParser(AttributeParser.OBJECT_PARSER)
                 .setAllowNull(allowNull);
     }
 
     /**
      * Utility method to return part of {@link ObjectTypeAttributeDefinition} for credential reference attribute.
      *
      * {@see CredentialReference#getAttributeDefinition}
-     * @param context operational context
-     * @param attributeDefinition attribute definition
-     * @param model model
+     * @param credentialReferenceValue value of credential reference attribute
      * @param name name of part to return (supported names: {@link #STORE} {@link #ALIAS} {@link #TYPE}
      *    {@link #CLEAR_TEXT}
      * @return value of part as {@link String}
      * @throws OperationFailedException when something goes wrong
      */
-    public static String credentialReferencePartAsStringIfDefined(OperationContext context, ObjectTypeAttributeDefinition attributeDefinition, ModelNode model, String name) throws OperationFailedException {
-        ModelNode value = attributeDefinition.resolveModelAttribute(context, model);
-        if (value.isDefined()) {
-            ModelNode namedNode = value.get(name);
-            if (namedNode != null && namedNode.isDefined()) {
-                return namedNode.asString();
-            }
-            return null;
+    public static String credentialReferencePartAsStringIfDefined(ModelNode credentialReferenceValue, String name) throws OperationFailedException {
+        assert !credentialReferenceValue.isDefined() : credentialReferenceValue;
+        ModelNode result = credentialReferenceValue.get(name);
+        if (result.isDefined()) {
+            return result.asString();
         }
         return null;
```

Comment:
Something seems odd here.

L148 asserts that credentialReferenceValue is undefined. So the test at L150 will always test false and the method will return null.

On L148 did you not intend for the '!' to be there?